lib/Conversion/ImportAIGER/ImportAIGER.cpp

```diff
@@ -0,0 +1,982 @@
+
+  if (isBinaryFormat) {
+    return parseAndGatesBinary();
+  }
+  return parseAndGatesASCII();
+}
+
+ParseResult AIGERParser::parseAndGatesASCII() {
+  auto startLoc = lexer.getCurrentLoc();
+  // Parse AND gate definitions: lhs rhs0 rhs1
+  for (unsigned i = 0; i < numAnds; ++i) {
+    unsigned lhs, rhs0, rhs1;
+    if (parseNumber(lhs) || parseNumber(rhs0) || parseNumber(rhs1) ||
+        parseNewLine())
+      return emitError(startLoc, "failed to parse AND gate definition");
+
+    LLVM_DEBUG(llvm::dbgs() << "AND Gate " << i << ": " << lhs << " = " << rhs0
+                            << " & " << rhs1 << "\n");
+
+    // Validate LHS (should be even and positive)
+    if (lhs % 2 != 0 || lhs == 0)
+      return emitError(startLoc, "invalid AND gate LHS literal");
+
+    // Validate literal bounds
+    if (lhs / 2 > maxVarIndex || rhs0 / 2 > maxVarIndex ||
+        rhs1 / 2 > maxVarIndex)
+      return emitError(startLoc,
+                       "AND gate literal exceeds maximum variable index");
+
+    andGateDefs.push_back({lhs, rhs0, rhs1, startLoc});
+  }
+
+  return success();
+}
+
+ParseResult AIGERParser::parseAndGatesBinary() {
+  // In binary format, AND gates are encoded with delta compression
+  // Each AND gate is encoded as: delta0 delta1
+  // where: rhs0 = lhs - delta0, rhs1 = rhs0 - delta1
+  auto startLoc = lexer.getCurrentLoc();
+
+  LLVM_DEBUG(llvm::dbgs() << "Starting binary AND gate parsing\n");
+
+  // First AND gate LHS starts after inputs and latches
+  // Variables are numbered: 1, 2, ..., maxVarIndex
+  // Literals are: 2, 4, 6, ..., 2*maxVarIndex
+  // Inputs: 2, 4, ..., 2*numInputs
+  // Latches: 2*(numInputs+1), 2*(numInputs+2), ..., 2*(numInputs+numLatches)
+  // AND gates: 2*(numInputs+numLatches+1), 2*(numInputs+numLatches+2), ...
+  auto currentLHS = 2 * (numInputs + numLatches + 1);
+
+  LLVM_DEBUG(llvm::dbgs() << "First AND gate LHS should be: " << currentLHS
+                          << "\n");
+
+  for (unsigned i = 0; i < numAnds; ++i) {
+    unsigned delta0, delta1;
+    if (parseBinaryNumber(delta0) || parseBinaryNumber(delta1))
+      return emitError("failed to parse binary AND gate deltas");
+
+    auto lhs = static_cast<int64_t>(currentLHS);
+
+    // Check for underflow before subtraction
+    if (delta0 > lhs || delta1 > (lhs - delta0)) {
+      LLVM_DEBUG(llvm::dbgs() << "Delta underflow: lhs=" << lhs << ", delta0="
+                              << delta0 << ", delta1=" << delta1 << "\n");
+      return emitError("invalid binary AND gate: delta causes underflow");
+    }
+
+    auto rhs0 = lhs - delta0;
+    auto rhs1 = rhs0 - delta1;
+
+    LLVM_DEBUG(llvm::dbgs() << "Binary AND Gate " << i << ": " << lhs << " = "
+                            << rhs0 << " & " << rhs1 << " (deltas: " << delta0
+                            << ", " << delta1 << ")\n");
+
+    // Validate the decoded literals
+    if (lhs % 2 != 0 || lhs == 0)
+      return emitError("invalid binary AND gate LHS literal");
```

Comment:
Can this ever happen? `auto currentLHS = 2 * (... + 1)` feels like it passes this check, and then `currentLHS` is only ever `+= 2`, which should maintain the property. :thinking: